src/middleware.rs

```diff
@@ -0,0 +1,108 @@
+use actix_web::{
+    body::EitherBody,
+    dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform},
+    Error, FromRequest, HttpMessage, HttpResponse,
+};
+use actix_web_httpauth::extractors::bearer::BearerAuth;
+use anyhow::{Context, Result};
+use jwt_simple::prelude::*;
+use log::error;
+use std::{
+    future::{ready, Future, Ready},
+    pin::Pin,
+    rc::Rc,
+};
+
+pub const TOKEN_EXPIRE_HOURES: u64 = 2;
+
+pub struct Auth;
+
+impl<S, B> Transform<S, ServiceRequest> for Auth
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B>>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = AuthMiddleware<S>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ready(Ok(AuthMiddleware {
+            service: Rc::new(service),
+        }))
+    }
+}
+
+pub struct AuthMiddleware<S> {
+    service: Rc<S>,
+}
+
+type LocalBoxFuture<T> = Pin<Box<dyn Future<Output = T> + 'static>>;
+
+impl<S, B> Service<ServiceRequest> for AuthMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B>>;
+    type Error = Error;
+    type Future = LocalBoxFuture<Result<Self::Response, Self::Error>>;
+
+    forward_ready!(service);
+
+    fn call(&self, mut req: ServiceRequest) -> Self::Future {
+        let service = Rc::clone(&self.service);
+
+        Box::pin(async move {
+            if str::starts_with(req.path(), "/action") {
+                let mut payload = req.take_payload().take();
+
+                let auth = match BearerAuth::from_request(req.request(), &mut payload).await {
+                    Ok(b) => b,
+                    Err(_) => {
+                        error!("No auth header");
+                        return Ok(unauthorized_error(req).map_into_right_body());
+                    }
+                };
+
+                match verify_token(auth) {
+                    Ok(_) => (),
+                    Err(e) => {
+                        error!("User not authorized {}", e);
+
+                        return Ok(unauthorized_error(req).map_into_right_body());
+                    }
+                }
+            }
+
+            let res = service.call(req).await?;
+            Ok(res.map_into_left_body())
```

Comment:
If I understand the code correctly I could use a route such as "https://somedomain/./action/..." and the check in 61 would fail?
I guess the best case would be to do the sensible business logic _only_ in the good case, i.e., (82-83) only in the good path (73). Rather being to strict and having a bug that causes the system to not work as intended than having a bug that causes authentication to be too loose :).